parser/src/ast/expression.rs

```diff
@@ -0,0 +1,719 @@
+    /// skips values for which the selector evaluates to zero (false).
+    ///
+    /// When the comprehension is used as a constraint, this field is only valid for
+    /// use in integrity constraints.
+    pub selector: Option<ScalarExpr>,
+}
+impl ListComprehension {
+    /// Creates a new list comprehension.
+    pub fn new(
+        span: SourceSpan,
+        body: ScalarExpr,
+        mut context: ComprehensionContext,
+        selector: Option<ScalarExpr>,
+    ) -> Self {
+        let bindings = context.iter().map(|(name, _)| name).copied().collect();
+        let iterables = context.drain(..).map(|(_, iterable)| iterable).collect();
+        Self {
+            span,
+            bindings,
+            iterables,
+            body: Box::new(body),
+            selector,
+        }
+    }
+}
+impl Eq for ListComprehension {}
+impl PartialEq for ListComprehension {
+    fn eq(&self, other: &Self) -> bool {
+        self.bindings == other.bindings
+            && self.iterables == other.iterables
+            && self.body == other.body
+            && self.selector == other.selector
+    }
+}
+impl fmt::Debug for ListComprehension {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ListComprehension")
+            .field("bindings", &self.bindings)
+            .field("iterables", &self.iterables)
+            .field("body", self.body.as_ref())
+            .field("selector", &self.selector)
+            .finish()
+    }
+}
+
+/// Represents a function call (either a pure function or an evaluator).
+///
+/// Calls are permitted in a scalar expression context, but arguments to the
+/// callee may be non-scalar expressions - it is expected that the callee produces
+/// a scalar in such contexts.
+///
+/// Calls to pure functions return scalar or non-scalar values, and so they may be
+/// used any place other scalar expressions are supported.
+///
+/// Calls to evaluator functions on the other hand, produce no values, and may only
+/// appear as the top-level expression of a [Statement::Enforce], e.g. `enf system_constraints([system_cols])`
+#[derive(Clone, Spanned)]
+pub struct Call {
+    #[span]
+    pub span: SourceSpan,
+    pub callee: ResolvableIdentifier,
+    pub args: Vec<Expr>,
+    /// Used to store the type produced by the callee if resolved
+    ///
+    /// If this is `None`, and the callee is resolved, then the callee is an
+    /// evaluator function. Otherwise, it means the type is unknown/unavailable.
+    pub ty: Option<Type>,
+}
+impl Call {
+    pub fn new(span: SourceSpan, callee: Identifier, args: Vec<Expr>) -> Self {
+        Self {
+            span,
+            callee: ResolvableIdentifier::Unresolved(NamespacedIdentifier::Function(callee)),
+            args,
+            ty: None,
+        }
+    }
+
+    /// Returns true if the callee is a builtin function, e.g. `sum`
+    #[inline]
+    pub fn is_builtin(&self) -> bool {
+        self.callee.is_builtin()
+    }
```

Comment:
This isn't allowed for evaluator function calls (maybe this is handled elsewhere, but we should at least note it here)

Reply:
I'm not sure I follow - you mean calling sum/prod from an evaluator?

Reply:
Ooh I think I misread this as caller instead of callee. Let's ignore this until I revisit :)